include/exadg/incompressible_navier_stokes/spatial_discretization/spatial_operator_base.cpp

```diff
@@ -1865,8 +1871,11 @@ SpatialOperatorBase<dim, Number>::local_interpolate_stress_bc_boundary_face(
         vector traction =
           param.density * (viscosity * (grad_u + transpose(grad_u)) * normal - p * normal);
 
-        integrator_u.submit_dof_value(traction, index);
+        auto * values = integrator_u.begin_values() + q;
+        for(unsigned int comp = 0; comp < integrator_u.n_components; ++comp)
+          values[comp * integrator_u.n_q_points] = traction[comp];
```

Comment:
I wonder whether we should add a comment here, maybe a reference to a class/function/data structure in deal.II providing some information or explanation. As mentioned in another comment, why should we have a guarantee that this will never change in deal.II?

By looking at this code only, one might argue that the `n_components` values of one quadrature point are not next to each other in the array `values`. For a very large array `values`, this would probably be inefficient (data locality). Depending on the background/perspective, this might look counter-intuitive to a user/developer of ExaDG, so a comment providing some reasoning could be helpful.

Finally, I wonder whether this functionality wouldn't actually call for a public member function of FEEvaluation/Integrator classes in deal.II? (interface e.g. like `write_to_values(integrator, q, data)`)

What do you think @dominiktassilostill @kronbichler ?

Reply:
Adding a comment and a reference to `submit_value` would certainly make things more clear. As the code is reused multiple times in the library now (and I think there are even more places where it is applicable), creating a separate function (with more in detail descriptions) for it definitely makes sense but I do not know if creating a function in `FEEvaluation` (or its base class) is necessary or not.

Reply:
> By looking at this code only, one might argue that the n_components values of one quadrature point are not next to each other in the array values. For a very large array values, this would probably be inefficient (data locality).

The reason the data fields are not next to each other is because the computation kernels to do the interpolation/integration (`FEEvaluation::evaluate` and `integrate`) are set up component by component, with the component being a run-time parameter. However, run-time parameters cannot be used as strides without making the code significantly slower. Note also that data locality is not at all a problem here, as there is a loop around it. Caches and associated hardware prefetchers are pretty good at handling access to multiple "streams". There is a corner case of cache conflict misses, and you might ask the question under which parameters for our variables these appear. I recommend you to read https://en.wikipedia.org/wiki/Cache_performance_measurement_and_metric and check the structure of set-associative caches that are the standard on CPU today. I'd be happy to discuss more about this and possible mitigation strategies if you find problematic cases.

> Finally, I wonder whether this functionality wouldn't actually call for a public member function of FEEvaluation/Integrator classes in deal.II? (interface e.g. like `write_to_values(integrator, q, data)`)

If you want to make a suggestion @nfehn I'd be happy to discuss it. I am not entirely sure if the present code is an important enough use case, as I will elaborate in a later message, but having the code at hand would definitely help the decision making.